Beta addresses/abi update and model changes

dwalintukan opened the pull request:
FYI: This was branched off `beta_master` branch before but I rebased off `master` now, which is why is has the `beta_master` changes as well.

frankobe commented on .gitignore:
```diff
@@ -57,4 +102,30 @@ typings/
+
+### Windows ###
+# Windows thumbnail cache files
+Thumbs.db
```
dunt think this is necessary

dwalintukan replied:
Think we should have it in there in case anyone who works on a windows machine commits something to the repo. Usually I setup all my gitignore.io files to include all OS's: Mac, Windows, Linux

frankobe commented on src/schema/index.js:
```diff
@@ -19,25 +19,29 @@ type Topic {
   token: String!
-  name: String!
-  options: [String!]!
+  name: String
```
`name` should not be `null`

dwalintukan replied:
fixed

frankobe commented on src/schema/index.js:
```diff
@@ -19,25 +19,29 @@ type Topic {
-  name: String!
-  options: [String!]!
+  name: String
+  options: [String!]
```
`options` should not be `null`

dwalintukan replied:
fixed

frankobe commented on src/schema/index.js:
```diff
@@ -19,25 +19,29 @@ type Topic {
   amounts: [String!]!
   resultIdx: Int
   blockNum: Int!
+  betStartBlock: Int
```
is this bet only?

dwalintukan replied:
changed to `startBlock`

frankobe commented on src/schema/index.js:
```diff
@@ -92,26 +97,25 @@ input VoteFilter {
 
 type Mutation {
   createTopic(
+    version: Int!
```
version can be optional when create.

dwalintukan replied:
removed. doesn't make sense to have it for the unconfirmed items.

dwalintukan replied:
do I need to make this optional in the `typeDefs` to allow for this?

frankobe replied:
lets leave mutation as it is because not really use it

frankobe commented on src/schema/index.js:
```diff
@@ -92,26 +97,25 @@ input VoteFilter {
   ): Topic
 
   createOracle(
+    version: Int!
```
version can be optional when create.

dwalintukan replied:
removed

frankobe commented on src/schema/index.js:
```diff
@@ -92,26 +97,25 @@ input VoteFilter {
   ): Oracle
 
   createVote(
+    version: Int!
```
version can be optional when create.

dwalintukan replied:
removed

frankobe commented on src/schema/index.js:
```diff
@@ -153,7 +162,12 @@ enum _TokenType {
 // Generate the schema object from your types definition.
-module.exports = makeExecutableSchema({typeDefs, resolvers});
\ No newline at end of file
+module.exports = makeExecutableSchema({ typeDefs, resolvers });
```
nit: newline at EOF

dwalintukan replied:
done

frankobe commented on src/schema/resolvers.js:
```diff
@@ -1,8 +1,38 @@
+    options.sort = [];
+
+    _.each(orderBy, (order) => {
+      options.sort.push([order.field, (order.direction === "ASC" ? 'asc' : 'desc')]);
```
according to https://github.com/louischatriot/nedb sort section, nedb supports `{field: 1/-1}`, not ASC, DESC

dwalintukan replied:
👍 removed

frankobe commented on src/schema/resolvers.js:
```diff
@@ -1,8 +1,38 @@
+ * @param  {Number} skip    skip number in paging; default value DEFAULT_SKIP_NUM
+ * @return {Object}         
+ */
+function buildQueryOptions(orderBy, limit, skip) {
```
cant find where this func is used in `resolver`

dwalintukan replied:
removed entire method

frankobe commented on src/schema/resolvers.js:
```diff
@@ -113,7 +145,7 @@ module.exports = {
 
       if (skip) {
-        cursor.skip(skip);
+        options.skip(skip);
```
cant find `options`, should be `cursor`

dwalintukan replied:
fixed

frankobe commented on src/sync/contracts.js:
```diff
@@ -1,34 +1,32 @@
-    FinalResultSet: "e1dd417dcbeb9c130709b6712f14e2b3362b509f72c1b86daa86da8173d92477",
-    WinningsWithdrawn: "64bd7c266edce1b240f0ed2697cdca2e2478fb1dbc18ec833f80cda28a34c029"
+    abi: [{"constant":false,"inputs":[{"name":"_resultIndex","type":"uint8"},{"name":"_sender","type":"address"},{"name":"_amount","type":"uint256"}],"name":"voteFromOracle","outputs":[{"name":"","type":"bool"}],"payable":false,"stateMutability":"nonpayable","type":"function"},{"constant":true,"inputs":[],"name":"totalBotValue","outputs":[{"name":"","type":"uint256"}],"payable":false,"stateMutabi ...
+    FinalResultSet: '14959b24f45a8f41b814b331ae09533db9d7e7962ca200e484f849a1fd1955aa',
```
signature should be dynamically calculated

dwalintukan replied:
can change later when you decide if we should hardcode the eventName here or inside call.

frankobe commented on src/sync/index.js:
```diff
@@ -351,7 +116,7 @@ async function sync(db){
       // sync first and then update in case update object in current batch
       Promise.all(updatePromises).then(() =>{
-        console.log('Updated OracleResult & FinalResult');
+        console.log('Updated OracleResult & FinalResult\n');
```
console.log should be auto newline, isnt it?

dwalintukan replied:
I added this to "batch" the logs together and make it more readable:
```
searchlog TopicCreated
71053 - 71073: Retrieved 0 entries from TopicCreated
searchlog OracleResultSet
71053 - 71073: Retrieved 0 entries from OracleResultSet
searchlog DecentralizedOracleCreated
71053 - 71073: Retrieved 0 entries from DecentralizedOracleCreated
searchlog OracleResultVoted
71053 - 71073: Retrieved 0 entries from OracleResultVoted
searchlog CentralizedOracleCreated
71053 - 71073: Retrieved 0 entries from CentralizedOracleCreated
Synced Topics & Oracles
searchlog FinalResultSet
71053 - 71073: Retrieved 0 entries from FinalResultSet
Updated OracleResult & FinalResult

searchlog TopicCreated
71053 - 71073: Retrieved 0 entries from TopicCreated
searchlog OracleResultSet
71053 - 71073: Retrieved 0 entries from OracleResultSet
searchlog DecentralizedOracleCreated
71053 - 71073: Retrieved 0 entries from DecentralizedOracleCreated
searchlog OracleResultVoted
71053 - 71073: Retrieved 0 entries from OracleResultVoted
searchlog CentralizedOracleCreated
71053 - 71073: Retrieved 0 entries from CentralizedOracleCreated
Synced Topics & Oracles
searchlog FinalResultSet
71053 - 71073: Retrieved 0 entries from FinalResultSet
Updated OracleResult & FinalResult
```

frankobe commented on src/sync/index.js:
```diff
@@ -83,259 +84,23 @@ async function sync(db){
-      Promise.all(createTopicPromises).then(()=>{
-        resolve()
-      });
+    var syncTopicCreatedPromise = new Promise(async (resolve) => { 
```
I prefer to delay this refactor as it requires manual testing. I dunt refactor it before because
1. func wont be reused, only mk it cleaner
2. multi insert db optimization should be done first

dwalintukan replied:
as discussed, going to parse TopicCreated before Oracle parsing to ensure topic is available to populate oracle info.

frankobe commented on src/sync/index.js:
```diff
@@ -83,259 +84,23 @@ async function sync(db){
-        resolve()
-      });
+    var syncCentralizedOracleCreatedPromise = new Promise(async (resolve) => { 
+      syncCentralizedOracleCreated(resolve, db, startBlock, endBlock, removeHexPrefix, oraclesNeedInfoUpdate);
```
1. `async` without `await`?
2. `resolve` as last parameter

dwalintukan replied:
seems fine without awaiting. i think the await is handled in the Promise.all => then()

frankobe commented on src/sync/index.js:
```diff
@@ -369,7 +134,10 @@ async function sync(db){
-  async function(){
+  async function() {
+    // Insert Oracle names and options
+    await updateOracleInfo(db, oraclesNeedInfoUpdate);
```
is this because Oracle doesnt hv `name` & `options` anymore?

dwalintukan replied:
correct. removed this and added the name and options when oracles are created.

frankobe commented on src/sync/index.js:
```diff
@@ -429,20 +456,67 @@ async function updateCentralizedOraclesPassedResultSetEndBlock(currentBlockChain
+    let oracle;
+    try {
+      oracle = await db.Oracles.findOne({ 
+        address: oracleAddress 
```
use `_id`, it is an indexed primary key

dwalintukan replied:
The `_id` field of the TopicEvent isn't in the Oracle objects.

frankobe commented on src/sync/index.js:
```diff
@@ -429,20 +456,67 @@ async function updateCentralizedOraclesPassedResultSetEndBlock(currentBlockChain
+    let topic;
+    try {
+      topic = await db.Topics.findOne({
+        address: oracle.topicAddress
```
we can pass topicAddress in array to save it one more query

dwalintukan replied:
this is fixed in the latest changes. I get the topicAddress from the parsed Oracle obj.

frankobe commented on src/sync/index.js:
```diff
@@ -369,7 +134,10 @@ async function sync(db){
-  async function(){
+  async function() {
+    // Insert Oracle names and options
+    await updateOracleInfo(db, oraclesNeedInfoUpdate);
```
I dunt think we should put insert after init sync is done, would like to populate entries into db asap

dwalintukan replied:
done as discussed

frankobe commented on src/sync/models/topic.js:
```diff
@@ -14,27 +14,24 @@ class Topic {
       _id: this.topicAddress,
+      version: this.version,
       address: this.topicAddress,
       txid: this.txid,
```
no more `creator`address?

dwalintukan replied:
yes removed for now since contract does not use this and no need to return.

frankobe commented on src/schema/index.js:
```diff
@@ -32,12 +33,15 @@ type Oracle {
   amounts: [String!]!
   resultIdx: Int
   blockNum: Int!
+  startBlock: Int
```
mark startBlock as `Int!` for consistency or filter in the future, if no value passed, use `blockNum`

frankobe commented on src/sync/models/centralizedOracle.js:
```diff
@@ -14,26 +14,27 @@ class CentralizedOracle {
     this.consensusThreshold = this.rawLog['_consensusThreshold'].toJSON();
   }
 
+  setTopicInfo(topic) {
```
this should be in synchroniser file as model is direct translation of blockchain event

dwalintukan replied:
done

frankobe commented on src/sync/models/centralizedOracle.js:
```diff
@@ -42,12 +43,14 @@ class CentralizedOracle {
       amounts: _.fill(Array(this.numOfResults), '0'),
       resultIdx: null,
       blockNum: this.blockNum,
+      startBlock: this.bettingStartBlock,
```
if none, set to blockNum

dwalintukan replied:
done for decentralized

frankobe replied:
so smart contract guarantee bettingStartBlock wont be null?

dwalintukan commented:
This branch has a bug with updating new blocks. See #62 for details.

dwalintukan commented:
Figured out the bug, it was not due to the beta changes I made, but this:
```
let block = await db.Blocks.cfind({}).sort({blockNum:-1}).limit(1).exec();
startBlock = Math.max(blocks.blockNum + 1, startBlock);
// This was returning NaN

// so (currentBlockChainHeight-startBlock) was not working, stalling the parsing of future blocks.
Math.ceil((currentBlockChainHeight-startBlock)/batchSize)
```

dwalintukan commented:
Oh I also, refactored the promise.all chaining. I thought this was related to the parsing new blocks bug, but it wasn't. So I refactored the promise chain to use all async/awaits. Tested it many times already and seems to be working as intended.

frankobe commented on src/sync/index.js:
```diff
@@ -412,13 +152,281 @@ async function sync(db){
+
+    if (!topic) {
+      console.error(`ERROR: could not find Topic ${address} in db`);
+      return null;
```
would rather throw exception to be caught or u hv to handle null check

frankobe replied:
done

frankobe commented on src/sync/index.js:
```diff
@@ -412,13 +152,281 @@ async function sync(db){
+          try {
+            let centralOracle = new CentralizedOracle(blockNum, txid, rawLog);
+            const topic = await fetchTopic(db, centralOracle.eventAddress);
+            centralOracle.name = topic.name;
```
miss null check

frankobe replied:
done through exception

frankobe commented on src/sync/index.js:
```diff
@@ -412,13 +152,281 @@ async function sync(db){
+            const topic = await fetchTopic(db, centralOracle.eventAddress);
+            centralOracle.name = topic.name;
+            centralOracle.options = topic.options;
+            centralOracle = centralOracle.translate();
```
translate() first and add `name`, `option` later

frankobe replied:
done

frankobe commented on src/sync/index.js:
```diff
@@ -412,13 +152,281 @@ async function sync(db){
+  let topic;
+  try {
+    topic = await db.Topics.findOne({
+      address: address
```
`_id: address` to use index

frankobe commented on src/sync/index.js:
```diff
@@ -412,13 +152,281 @@ async function sync(db){
+          try {
+            var topic = new Topic(blockNum, txid, rawLog).translate();
+            await db.Topics.insert(topic);
+            resolve();
```
no `resolve` define

frankobe commented on src/sync/index.js:
```diff
@@ -412,13 +152,281 @@ async function sync(db){
+            centralOracle = centralOracle.translate();
+
+            await db.Oracles.insert(centralOracle);
+            resolve();
```
no resolve define

frankobe commented on src/sync/index.js:
```diff
@@ -412,13 +152,281 @@ async function sync(db){
+    });
+  });
+
+  Promise.all(createTopicPromises);
```
missing `await`?

frankobe replied:
added
